core/src/ldml/ldml_processor.cpp

```diff
@@ -336,5 +352,76 @@ km_kbp_status ldml_processor::validate() const {
   return _valid ? KM_KBP_STATUS_OK : KM_KBP_STATUS_INVALID_KEYBOARD;
 }
 
+void
+ldml_processor::emit_text(km_kbp_state *state, const std::u16string &str) {
+  const std::u32string str32 = kmx::u16string_to_u32string(str);
+  emit_text(state, str32);
+}
+
+void
+ldml_processor::emit_text(km_kbp_state *state, const std::u32string &str) {
+  for (auto it = str.begin(); it < str.end(); it++) {
+    const auto ch = *it;
+    // If we are at the start of a sequence:
+    if (ch == LDML_UC_SENTINEL) {
+      it++; // consume LDML_UC_SENTINEL
+      // TODO-LDML: Might assert if a malformed sequence is included- "should not happen"?
+      assert(it < str.end());
+      // verify that the next char is LDML_MARKER_CODE
+      assert(*it == LDML_MARKER_CODE);
+      it++; // consume LDML_MARKER_CODE
+      assert(it < str.end());
+      const auto marker_no = *it;
+      assert(marker_no >= LDML_MARKER_MIN_INDEX);
+      assert(marker_no <= LDML_MARKER_ANY_INDEX);
+      emit_marker(state, marker_no);
+    } else {
+      emit_text(state, ch);
+    }
+  }
+}
+
+void
+ldml_processor::emit_text(km_kbp_state *state, km_kbp_usv ch) {
+  assert(ch != LDML_UC_SENTINEL);
+  state->context().push_character(ch);
+  state->actions().push_character(ch);
+}
+
+void ldml_processor::emit_marker(km_kbp_state *state, KMX_DWORD marker_no) {
+    // OK, push the marker
+    state->actions().push_marker(marker_no);
+    state->context().push_marker(marker_no);
+}
+
+size_t
+ldml_processor::context_to_string(km_kbp_state *state, std::u32string &str) {
```

Comment:
Should we empty str before we start?

Reply:
we could, but i thought of it as an append.  might be safer as you propose

Reply:
That feels too side-effecty, should be `append_context_to_string` in that case.